### PR TITLE
Create Work Item Feature (ET)

### PR DESCRIPTION
**Pull request by t-emtenc**
UI Mock completed for the CreateWorkItem page. No REST API calls incorporated yet.

### CODE REVIEW

File: App/Reducers/ActionsET.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
+/**
+ * enum for Actions
+ * @type {enum}
+ */
+export enum ACTION {STAGE, GEAR, WORKITEMTYPE, NEWTITLE, NEWDESCRIPTION, ADDASATTACHMENT, FOLLOWSTATE, SAVE, POSTCREATION}
```
Comment (aygarg) [inline App/Reducers/ActionsET.ts]: Please add every item of the enum in a new line.

File: App/Reducers/ActionsET.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+/**
+ * enum for the stages in the Follow Process
+ * @type {enum}
+ */
+export enum FollowStateTypes {Followed, Unfollowed, Request}
+/**
+ * enum for Stage of the edit process of the work item
+ * @type {enum}
+ */
+export enum Stage {New, Draft, Saved}
```
Comment (aygarg) [inline App/Reducers/ActionsET.ts]: Give a blank line between two definitions

File: App/Reducers/ActionsET.ts [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+     * The id the newly created VSTS work item
+     * @type {string}
+     */
+  id: string;
+}
+  /**
+   * Handles update of the stage
+   * @param {Stage} stage
+   */
+export function changeStage (stage: Stage): IStageAction {
```
Comment (aygarg) [inline App/Reducers/ActionsET.ts]: Can use use update instead of using change everywhere.

File: App/Reducers/ReducersET.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  title : '',
+  workItemType: 'Bug',
+};
+
+  /**
+   * Handles changing the value of the fields with each action
+   * @param {IWorkItem} state
+   * @param {any} action
+   */
+function changeFields(state: IWorkItem = initalState, action: any): IWorkItem {
```
Comment (aygarg) [inline App/Reducers/ReducersET.ts]: This is a reducer then call it workItemReducer instead of changeFields

File: App/Reducers/ReducersET.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
+import { Reducer, combineReducers } from 'redux';
+import { ACTION, FollowStateTypes, Stage } from './ActionsET';
+ /**
+  * Represents the part of the state for the new WorkItem
+  * @interface IWorkItem
+  */
+export interface IWorkItem {
```
Comment (aygarg) [inline App/Reducers/ReducersET.ts:7]: Don't you need account, project and area here?

File: App/VSTS/CreateWorkItem.tsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import { Gear } from './Gear';
+
+/**
+ * Renders all components of the Create page
+ * @class { CreateWorkItem }
+ */
+
+export class CreateWorkItem extends React.Component<{}, {}> {
+
+  public isReady: boolean; // set to false
```
Comment (aygarg) [inline App/VSTS/CreateWorkItem.tsx:17]: Why are we using isReady? is this a temp workaround till you integrate with Miranda's work? then its fine.

File: App/VSTS/Description.tsx [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+/**
+ * Maps elements of the state to properties
+ * @returns {IDescriptionProps}
+ * @param {any} state
+ */
+function mapStateToProps (state: any): IDescriptionProps  {
+  return {addAsAttachment: state.workItem.addAsAttachment, description: state.workItem.description} ;
+   }
+
+@connect (mapStateToProps)
```
Comment (aygarg) [inline App/VSTS/Description.tsx:37]: Keep it just before the component definition after the comment for it.

File: App/VSTS/Description.tsx [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+  this.props.dispatch(changeDescription (event.target.value));
+}
+
+/**
+ * Dispatches the action to update the addAsAttachment and description values in the store
+ * @ returns {void}
+ */
+public handlechangeAddAsAttachment (): void {
+  if (!this.props.addAsAttachment) {
+      this.props.dispatch(changeDescription('For more details, please refer to the attached email thread. ' + this.props.description));
```
Comment (aygarg) [inline App/VSTS/Description.tsx]: Use a const string for this description.

File: App/VSTS/Save.tsx [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+ * @type {boolean}
+ */
+   public isDisabled: boolean = false;
+/**
+ * Dispatches the action to change the Stage and make the REST call to create the work item
+ * @returns {void}
+ */
+   public handleSave(): void {
+      this.props.dispatch(changeStage(Stage.Saved));
+      Rest.createWorkItem ('t-emtenc@microsoft.com', 'o365exchange', 'Outlook Services/Ecosystem - Ext VSTS', 'Bug',
```
Comment (aygarg) [inline App/VSTS/Save.tsx]: Why are you not reading this from Store?

File: App/VSTS/Title.tsx [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    /**
+     * title of the work item
+     * @type {string}
+     */
+    title?: string;
+    /**
+     * Flag to signal the stage the user is on: New if no edits are make, Draft if edits were made, Saved if the user created the work item
+     * @type {Stage}
+     */
+    stage?: Stage;
```
Comment (aygarg) [inline App/VSTS/Title.tsx]: Why do you have stage inside title? Don't you have one at work item level.

File: App/VSTS/Title.tsx [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    width: '250px',
+  };
+/**
+ * Gets the normalizedSubject from Office and depending on the Stage, dispatches an action to update the value of title in store
+ */
+  let normalizedSubject: string = Office.context.mailbox.item.normalizedSubject;
+  let currentTitle: string = this.props.title;
+  if (currentTitle === '' && this.props.stage === Stage.New) {
+      this.props.dispatch(changeTitle (normalizedSubject));
+      this.props.dispatch(changeStage (Stage.Draft));
```
Comment (aygarg) [inline App/VSTS/Title.tsx]: mark it as draft in the same action as changeTitle instead of dispatching one more action to set it to draft.

File: Routes/Authenticate.js [hunk 12]
```diff
@@ -162,8 +162,8 @@ router.refreshToken = function (state, res) {
     assertion: state.refresh,
```
Comment (aygarg) [inline Routes/Authenticate.js:163]: Please check that all of you are not changing the same files.

File: app/Index.tsx [hunk 13]
```diff
@@ -1,12 +1,13 @@
 /// <reference path="../typings/tsd.d.ts" />
```
Comment (aygarg) [inline app/Index.tsx:1]: Don't checkin change to this file I guess. Otherwise all of you will have conflict.

File: app/VSTS/LogInPage.tsx [hunk 14]
```diff
@@ -1,76 +0,0 @@
-import * as React from 'react';
```
Comment (aygarg) [inline app/VSTS/LogInPage.tsx]: Why are you deleting this file?

File: app/VSTS/VSTS.tsx [hunk 15]
```diff
@@ -1,39 +1,42 @@
 /// <reference path="../../office.d.ts" />
```
Comment (aygarg) [inline app/VSTS/VSTS.tsx]: you should not change the vsts file. I mean don't include it in your pull request.

File: app/workItemModel.ts [hunk 16]
```diff
@@ -7,5 +7,4 @@ export interface IField {
 export interface IWorkItemState {
```
Comment (aygarg) [inline app/workItemModel.ts:7]: are you still using this?

File: config/webpack.prod.js [hunk 17]
```diff
@@ -1,47 +1,51 @@
 var path = require('path');
-var webpack = require('webpack');
+ var webpack = require('webpack');
 
 var APP_DIR = path.join(__dirname, '..', 'app');
 
 module.exports = {
-  devtool: 'source-map',
```
Comment (aygarg) [inline config/webpack.prod.js:7]: what are these changes for?

File: debug.js [hunk 18]
```diff
@@ -1 +1,2 @@
-module.exports = false;
\ No newline at end of file
+module.exports = false;
```
Comment (aygarg) [inline debug.js:1]: again why is this change in this pull request?

File: package.json [hunk 19]
```diff
@@ -44,10 +44,12 @@
     "react-dom": "^0.14.7",
     "react-redux": "^4.1.2",
     "react-router": "^2.4.1",
+    "react-select": "^1.0.0",
     "react-transform-catch-errors": "^1.0.2",
     "react-transform-hmr": "^1.0.2",
     "redbox-react": "^1.2.2",
     "redux": "^3.1.5",
+    "redux-form": "^5.3.1",
```
Comment (aygarg) [inline package.json:55]: do you need redux form ?

File: tsd.json [hunk 20]
```diff
@@ -15,7 +15,7 @@
       "commit": "6052959591c9bc6f8463c58220965c7030ba934d"
     },
     "react/react.d.ts": {
-      "commit": "6052959591c9bc6f8463c58220965c7030ba934d"
+      "commit": "c6c068935a58aa58087a309740bcf933600140b8"
```
Comment (aygarg) [inline tsd.json:18]: why this change?

File: typings/react/react.d.ts [hunk 21]
```diff
@@ -150,6 +150,7 @@ declare namespace __React {
     var DOM: ReactDOM;
     var PropTypes: ReactPropTypes;
```
Comment (aygarg) [inline typings/react/react.d.ts:151]: Why do you have these changes?

File: App/Reducers/ActionsET.ts (hunk 1 above)
Comment (t-emtenc) [inline App/Reducers/ActionsET.ts]: Resolved

File: App/Reducers/ActionsET.ts (hunk 2 above)
Comment (t-emtenc) [inline App/Reducers/ActionsET.ts]: Resolved

File: App/Reducers/ActionsET.ts (hunk 3 above)
Comment (t-emtenc) [inline App/Reducers/ActionsET.ts]: Resolved

File: App/Reducers/ReducersET.ts (hunk 4 above)
Comment (t-emtenc) [inline App/Reducers/ReducersET.ts]: Resolved

File: App/Reducers/ReducersET.ts (hunk 5 above)
Comment (t-emtenc) [inline App/Reducers/ReducersET.ts:7]: I am going to use the account, project, and area from Miranda's part of the store.

File: App/VSTS/CreateWorkItem.tsx (hunk 6 above)
Comment (t-emtenc) [inline App/VSTS/CreateWorkItem.tsx:17]: This is a temp workaround until I add Miranda's part.

File: App/VSTS/Description.tsx (hunk 7 above)
Comment (t-emtenc) [inline App/VSTS/Description.tsx:37]: Resolved

File: App/VSTS/Description.tsx (hunk 8 above)
Comment (t-emtenc) [inline App/VSTS/Description.tsx]: Are you saying use a const string instead of the event: any?

File: App/VSTS/Save.tsx (hunk 9 above)
Comment (t-emtenc) [inline App/VSTS/Save.tsx]: These values are hard-coded until I get the account, project, and area from Miranda's part of the store.

File: App/VSTS/Title.tsx (hunk 11 above)
Comment (t-emtenc) [inline App/VSTS/Title.tsx]: I wanted to separate the stage from the title component so if a "back" button was instituted later, you can change the stage without updating the title.

File: Routes/Authenticate.js (hunk 12 above)
Comment (t-emtenc) [inline Routes/Authenticate.js:163]: Andrew and I synced up to make sure that all my files match master.

File: app/VSTS/LogInPage.tsx (hunk 14 above)
Comment (t-emtenc) [inline app/VSTS/LogInPage.tsx]: This is Miranda's part, so she will merge her updated file.

File: app/workItemModel.ts (hunk 16 above)
Comment (t-emtenc) [inline app/workItemModel.ts:7]: I think this is Andrew's code.

File: config/webpack.prod.js (hunk 17 above)
Comment (t-emtenc) [inline config/webpack.prod.js:7]: These changes were to get the React-Select style working.

File: package.json (hunk 19 above)
Comment (t-emtenc) [inline package.json:55]: Resolved.

File: tsd.json (hunk 20 above)
Comment (t-emtenc) [inline tsd.json:18]: I don't remember changing this. I'll investigate where it came from.

File: typings/react/react.d.ts (hunk 21 above)
Comment (t-emtenc) [inline typings/react/react.d.ts:151]: I don't remember changing this either. I'll investigate where they both came from.

File: App/Reducers/ReducersET.ts (hunk 5 above)
Comment (aygarg) [inline App/Reducers/ReducersET.ts:7]: Ok.

File: debug.js (hunk 18 above)
Comment (annich-MS) [inline debug.js:1]: This is fine, this is newline at end of file (formatting fix)

File: config/webpack.prod.js [hunk 22]
```diff
@@ -1,47 +1,51 @@
...
+     root: [path.resolve('../app')],
+     extensions: ['', '.jsx', '.js', '.tsx', '.ts', '.css']
+   },
+   tslint: {
+     emitErrors: false,
+     failOnHint: false
+   }
+ }
+
+
```
Comment (annich-MS) [inline config/webpack.prod.js]: Extra Lines
